Test Improvements

MaartenBent opened the pull request:
This PR fixes several issues with the tests.

- Copy the data files and rc file when using CMake to generate the test projects.
Add headers to source file list so they are shown in IDE.
- Check for any error code when the tests finish under AppVeyor. It can return different error codes, e.g. for a segfault (see comments below).
- Generalize printf tests for pointers and results with an exponent. Compilers return slightly different values, e.g. capitalized pointer addresses and mantissas with length 2 or 3. The pre-processor test (setting `wxUSING_VC_CRT_IO`) for the correct result is not valid anymore, e.g. recent MinGW compilers use a mantissa with 2 zeros while 3 were expected. Simplify by accepting two valid results.
- Use the same `ctest` commands for AppVeyor and Travis. Don't run gui tests on Travis, these fail with `Error: Unable to initialize GTK+, is DISPLAY set properly?`.
- Prevent a segfault in the vararg test with MinGW. The segmentation fault is caused by `swscanf` with `__USE_MINGW_ANSI_STDIO=1`. See https://sourceforge.net/p/mingw-w64/mailman/message/36118530/ and comments below for a minimal example.

MaartenBent committed:
CMake: copy data files of test projects

Add headers to source files so they are included in IDE projects.

MaartenBent committed:
Check all error codes of AppVeyor tests

A test might return a different error code than 1,
for example when a segfault occurs.

MaartenBent committed:
Use same ctest command on Travis and AppVeyor

Do not run the GUI tests on Travis, these fail with:
Error: Unable to initialize GTK+, is DISPLAY set properly?

MaartenBent committed:
Prevent segfault in vararg test with MinGW

The segmentation fault is caused by swscanf with __USE_MINGW_ANSI_STDIO=1.
See https://sourceforge.net/p/mingw-w64/mailman/message/36118530/

vadz commented on build/tools/appveyor-test.bat:
```diff
@@ -9,27 +9,27 @@ goto %TOOLSET%
 PATH=C:\projects\wxwidgets\lib\vc_x64_dll;%PATH%
 .\vc_x64_mswudll\test.exe
-if errorlevel 1 goto :error
+if %errorlevel% NEQ 0 goto :error
```
Just out of curiosity, what's the difference between the old and the new versions? AFAIK they should do the same thing, don't they?

MaartenBent replied:
When I got a segfault, it returned `-1073741819`, not `1`. Currently the segfault is not correctly detected, see for example [this test](https://ci.appveyor.com/project/wxWidgets/wxwidgets/build/3387/job/j5a4kaet7bch95nd#L1771). It just stops (no finished message or summary) and continues with the next test. I hope this will detect it.

MaartenBent replied:
[this test](https://ci.appveyor.com/project/wxWidgets/wxwidgets/build/3408/job/61opifr6q26tjf13#L1758) shows the new version does detect the error.

vadz commented on tests/strings/strings.cpp:
```diff
@@ -845,17 +845,12 @@ void StringTestCase::FromDouble()
     } testData[] =
     {
         { 1.23,             -1, "1.23" },
-        // All MSVC versions until MSVC 14 used 3 digits for the exponent
```
Why did this have to be changed?

MaartenBent replied:
It fails with MinGW, similar as [this](https://ci.appveyor.com/project/wxWidgets/wxwidgets/build/3387/job/j5a4kaet7bch95nd#L1770).

vadz commented on tests/strings/vararg.cpp:
```diff
@@ -200,18 +200,23 @@ void VarArgTestCase::Sscanf()
     CPPUNIT_ASSERT( wxString(str) == "test" );
 
+#ifndef __MINGW32__
+    // disable this test on mingw to prevent segmentation fault:
```
Should this be always done or just for some value of `__USE_MINGW_ANSI_STDIO`?

MaartenBent replied:
It seems only with `__USE_MINGW_ANSI_STDIO=1`. For example, the code below works. But when enabling `#include <string>` it crashes.

```c++
//#include <string>
#include <wchar.h>

#define VALUE_TO_STRING(x) #x
#define VALUE(x) VALUE_TO_STRING(x)
#define VAR_NAME_VALUE(var) #var "="  VALUE(var)

int main()
{
#pragma message(VAR_NAME_VALUE(__USE_MINGW_ANSI_STDIO))

  wchar_t sentence [] = L"42 test";
  wchar_t str [20];
  int i;

  swscanf (sentence,L"%d %s",&i,str);
  wprintf (L"%s -> %d\n",str,i);

  return 0;
}

```

MaartenBent replied:
Unfortunately something is pulling in stl headers, probably Catch2.

vadz commented:
The only part I don't feel 100% comfortable about is allowing 2 valid results for printf tests. AFAIK for the given compiler (and version), only one of them is valid, so it looks rather hackish to accept both of them.

It's probably not a big deal in practice and if you think it's better like this (but why?) or that the old version couldn't be fixed to work with MinGW (but again I wonder why couldn't we check for its version too?), let's apply it like this. But ideally, I think, we'd accept only the single correct result and not both possibilities.

MaartenBent committed:
Fix printf tests for recent MinGW versions

The default mantissa length was changed from 3 to 2 in MinGW 5.0.4.
Fix the test for printing pointer addresses in 64bit MinGW.

MaartenBent commented:
I added a version check for MinGW and define `wxDEFAULT_MANTISSA_SIZE_3` for all known cases. Printing exponents was changed in [MinGW 5.0.4](https://sourceforge.net/p/mingw-w64/mailman/message/36333746/).

vadz commented:
Great, thanks for the update! I'll merge it when the CI builds pass, although it might not happen immediately as I'll have little/bad network connectivity for a few days.

vadz commented:
Merged now, thanks again for working on this!
